src/esi/Expression.cc

```diff
@@ -280,6 +284,7 @@ evalor(stackmember * stack, int *depth, int whereAmI, stackmember * candidate)
 
     srv.precedence = 1;
 
+    Must2(*depth < STACK_DEPTH_LIMIT-1, "expression too complex");
```

Comment:
Please move the `stack` addition code (i.e. this Must2() and the increment/assignment below it) into a new dedicated AddToStack() or some such function to avoid code duplication. Same for all other similar places in the PR, of course.

Ideally, we should refactor expression storage instead, but such refactoring probably goes beyond what you want to do in this PR.

Reply:
Indeed. I have added stackpush() to match the stackpop() already present. Replacing the stack with std::stack is left for later.